src/unix/lwt_unix_unix.h

```diff
@@ -1824,26 +1813,28 @@ struct job_readdir_n {
     DIR *dir;
     long count;
     int error_code;
-    struct dirent entries[];
+    char *entries[];
 };
 
 static void worker_readdir_n(struct job_readdir_n *job)
 {
     long i;
     for (i = 0; i < job->count; i++) {
-        struct dirent *ptr;
-        int result = readdir_r(job->dir, &job->entries[i], &ptr);
+        errno = 0;
+        struct dirent *entry = readdir(job->dir);
 
         /* An error happened. */
-        if (result != 0) {
-            job->error_code = result;
+        if (entry == NULL && errno != 0) {
+            job->error_code = errno;
             return;
         }
 
         /* End of directory reached */
-        if (ptr == NULL) break;
-    }
+        if (entry == NULL && errno == 0) break;
 
+        /* All is good */
+        job->entries[i] = strdup(entry->d_name);
```

Comment:
Should check the result of this for `NULL`, and probably set a custom error code (`ENOMEM`?) and break out of the loop to ensure a valid count of entries previously allocated successfully. See other comments on the count and `free`.